### PR TITLE
Modernise importer.py and type ModuleDetails

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
Modernise importer.py and type ModuleDetails
```

## Additional Context

Minor refactoring extracted from https://github.com/canonical/cloud-init/pull/1597

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by aciba90** [comment]
Force pushed to resolve conflicts.

File: cloudinit/importer.py [hunk 1]
```diff
@@ -8,16 +8,34 @@
...
+def _count_attrs(
+    module_name: str, attrs: Optional[Sequence[str]] = None
+) -> int:
+    found_attrs = 0
+    if not attrs:
+        return found_attrs
+    mod = importlib.import_module(module_name)
+    for attr in attrs:
+        if hasattr(mod, attr):
+            found_attrs += 1
```
Comment (holmanb) [inline cloudinit/importer.py:29]: Is the number of found attributes really the important thing to test in `find_module()`? In other words, if the module has every required attribute, why should it not be imported?

If the module has every required attribute, but has additional attributes that are not included in the required_attrs argument, won't this counting test cause import to fail? I agree that this looks functionally similar to what we're already doing, I wonder whether it is unnecessarily brittle. If I add an attribute to a module and now the runtime importer fails, I'd be confused - and having to dig into the magical import machinery to figure out what's going on is less than ideal for anyone writing a module.

**Review by holmanb — APPROVED** [review]
I left a question inline, but since it appears to match current behavior I'm not going to block this work on my comment/question. Let me know if you want to do anything further, otherwise I'll merge. Thanks for this!

**Comment by aciba90** [comment]
> I left a question inline, but since it appears to match current behavior I'm not going to block this work on my comment/question. Let me know if you want to do anything further, otherwise I'll merge. Thanks for this!

Thanks, @holmanb, for the review. I have answered you and added a comment within the code to make it 99% clear.

**Pull request merged**